### PR TITLE
GPU: Zero-init handling

### LINKED ISSUE #10785: GPU: Reasonable zero-init defaults
This boils down to four tasks.

- 0 should be an invalid value on specific enums, like formats.
- 0 should be a "reasonable default" choice on all other enums.
- Fields where 0 is not a reasonable default should have an "enable" bool to activate them.
- Fields where 0 is invalid should assert in the debug layer.

### PR DESCRIPTION
**Pull request by thatcosmonaut**
Resolves #10785

### CODE REVIEW

File: src/gpu/SDL_gpu.c [hunk 1]
```diff
@@ -80,11 +80,17 @@
     }
 
 #define CHECK_TEXTUREFORMAT_ENUM_INVALID(format, retval)     \
-    if (format >= SDL_GPU_TEXTUREFORMAT_MAX) {               \
+    if (format == 0 || format >= SDL_GPU_TEXTUREFORMAT_MAX) {               \
```
Comment (TheSpydog) [inline src/gpu/SDL_gpu.c:83]: We should probably use SDL_GPU_TEXTUREFORMAT_INVALID rather than 0

File: src/gpu/SDL_gpu.c [hunk 2]
```diff
@@ -80,11 +80,17 @@
...
         SDL_assert_release(!"Invalid texture format enum!"); \
         return retval;                                       \
     }
 
+#define CHECK_VERTEXELEMENTFORMAT_ENUM_INVALID(format, retval)       \
+    if (format == 0 || format >= SDL_GPU_VERTEXELEMENTFORMAT_MAX) {  \
```
Comment (TheSpydog) [inline src/gpu/SDL_gpu.c:89]: ditto

File: src/gpu/vulkan/SDL_gpu_vulkan.c [hunk 3]
```diff
@@ -337,6 +338,7 @@ static VkComponentMapping SwapchainCompositionSwizzle[] = {
 };
 
 static VkFormat SDLToVK_VertexFormat[] = {
+    VK_FORMAT_UNDEFINED,           // UNKNOWN
```
Comment (TheSpydog) [inline src/gpu/vulkan/SDL_gpu_vulkan.c:341]: I think this is supposed to be INVALID?

File: include/SDL3/SDL_gpu.h [hunk 4]
```diff
@@ -586,12 +586,13 @@ typedef enum SDL_GPUShaderStage
  */
 typedef Uint32 SDL_GPUShaderFormat;
 
-#define SDL_GPU_SHADERFORMAT_PRIVATE  (1u << 0) /**< Shaders for NDA'd platforms. */
-#define SDL_GPU_SHADERFORMAT_SPIRV    (1u << 1) /**< SPIR-V shaders for Vulkan. */
-#define SDL_GPU_SHADERFORMAT_DXBC     (1u << 2) /**< DXBC SM5_0 shaders for D3D11. */
-#define SDL_GPU_SHADERFORMAT_DXIL     (1u << 3) /**< DXIL shaders for D3D12. */
-#define SDL_GPU_SHADERFORMAT_MSL      (1u << 4) /**< MSL shaders for Metal. */
-#define SDL_GPU_SHADERFORMAT_METALLIB (1u << 5) /**< Precompiled metallib shaders for Metal. */
+#define SDL_GPU_SHADERFORMAT_INVALID  (1u << 0)
```
Comment (slouken) [inline include/SDL3/SDL_gpu.h:589]: Do you want SDL_GPU_SHADERFORMAT_INVALID to be 0? This will be 1 as-is.

File: src/gpu/SDL_gpu.c [hunk 5]
```diff
@@ -1,4 +1,4 @@
-/*
+﻿/*
```
Comment (slouken) [inline src/gpu/SDL_gpu.c:1]: BOM?

File: src/gpu/SDL_gpu.c [hunk 6]
```diff
@@ -643,6 +670,34 @@ SDL_GPUGraphicsPipeline *SDL_CreateGPUGraphicsPipeline(
...
             }
+            if (graphicsPipelineCreateInfo->target_info.color_target_descriptions[i].blend_state.enable_blend) {
+                const SDL_GPUColorTargetBlendState *blend_state = &graphicsPipelineCreateInfo->target_info.color_target_descriptions[i].blend_state;
+                CHECK_BLENDFACTOR_ENUM_INVALID(blend_state->src_color_blendfactor, NULL)
+                CHECK_BLENDFACTOR_ENUM_INVALID(blend_state->dst_color_blendfactor, NULL)
+                CHECK_BLENDOP_ENUM_INVALID(blend_state->color_blend_op, NULL)
+                CHECK_BLENDFACTOR_ENUM_INVALID(blend_state->src_alpha_blendfactor, NULL)
+                CHECK_BLENDFACTOR_ENUM_INVALID(blend_state->dst_alpha_blendfactor, NULL)
+                CHECK_BLENDOP_ENUM_INVALID(blend_state->alpha_blend_op, NULL)
+                if (blend_state->src_color_blendfactor == SDL_GPU_BLENDFACTOR_INVALID) {
```
Comment (slouken) [inline src/gpu/SDL_gpu.c:681]: Is this redundant with the check above?

Comment (thatcosmonaut) [reply src/gpu/SDL_gpu.c:681]: Yep, forgot to delete those...

File: include/SDL3/SDL_gpu.h (hunk 4 above)
Comment (thatcosmonaut) [reply include/SDL3/SDL_gpu.h:589]: Yeah that should be 0

File: include/SDL3/SDL_gpu.h [hunk 7]
```diff
@@ -1251,17 +1258,17 @@ typedef struct SDL_GPUStencilOpState
...
-    Uint8 padding2;
-    Uint8 padding3;
     SDL_GPUBlendFactor src_color_blendfactor;     /**< The value to be multiplied by the source RGB value. */
     SDL_GPUBlendFactor dst_color_blendfactor;     /**< The value to be multiplied by the destination RGB value. */
     SDL_GPUBlendOp color_blend_op;                /**< The blend operation for the RGB components. */
     SDL_GPUBlendFactor src_alpha_blendfactor;     /**< The value to be multiplied by the source alpha. */
     SDL_GPUBlendFactor dst_alpha_blendfactor;     /**< The value to be multiplied by the destination alpha. */
     SDL_GPUBlendOp alpha_blend_op;                /**< The blend operation for the alpha component. */
-    SDL_GPUColorComponentFlags color_write_mask;  /**< A bitmask specifying which of the RGBA components are enabled for writing. */
+    SDL_GPUColorComponentFlags color_write_mask;  /**< A bitmask specifying which of the RGBA components are enabled for writing. Writes to all channels if color_write_mask is SDL_FALSE. */
```
Comment (TheSpydog) [inline include/SDL3/SDL_gpu.h:1267]: `color_write_mask` in the comment needs to be `enable_color_write_mask`

File: include/SDL3/SDL_gpu.h [hunk 8]
```diff
@@ -250,10 +250,10 @@ typedef struct SDL_GPUFence SDL_GPUFence;
  */
 typedef enum SDL_GPUPrimitiveType
 {
+    SDL_GPU_PRIMITIVETYPE_TRIANGLELIST,  /**< A series of separate triangles. */
     SDL_GPU_PRIMITIVETYPE_POINTLIST,     /**< A series of separate points. */
     SDL_GPU_PRIMITIVETYPE_LINELIST,      /**< A series of separate lines. */
     SDL_GPU_PRIMITIVETYPE_LINESTRIP,     /**< A series of connected lines. */
-    SDL_GPU_PRIMITIVETYPE_TRIANGLELIST,  /**< A series of separate triangles. */
     SDL_GPU_PRIMITIVETYPE_TRIANGLESTRIP  /**< A series of connected triangles. */
```
Comment (TheSpydog) [inline include/SDL3/SDL_gpu.h:257]: Let's maybe swap this and POINTLIST so the triangle primitive types are together?

File: src/gpu/SDL_gpu.c [hunk 9]
```diff
@@ -80,19 +80,49 @@
...
 
 #define CHECK_TEXTUREFORMAT_ENUM_INVALID(format, retval)     \
-    if (format >= SDL_GPU_TEXTUREFORMAT_MAX) {               \
+    if (format == SDL_GPU_TEXTUREFORMAT_INVALID || format >= SDL_GPU_TEXTUREFORMAT_MAX_ENUM_VALUE) {               \
         SDL_assert_release(!"Invalid texture format enum!"); \
         return retval;                                       \
     }
 
+#define CHECK_VERTEXELEMENTFORMAT_ENUM_INVALID(format, retval)       \
+    if (format == SDL_GPU_VERTEXELEMENTFORMAT_INVALID || format >= SDL_GPU_VERTEXELEMENTFORMAT_MAX_ENUM_VALUE) {  \
```
Comment (TheSpydog) [inline src/gpu/SDL_gpu.c:89]: Should these INVALID checks be <= instead of == to handle negative numbers?

File: src/gpu/d3d11/SDL_gpu_d3d11.c [hunk 10]
```diff
@@ -1255,6 +1268,8 @@ static ID3D11BlendState *D3D11_INTERNAL_FetchBlendState(
     D3D11_BLEND_DESC blendDesc;
     HRESULT res;
 
+
```
Comment (TheSpydog) [inline src/gpu/d3d11/SDL_gpu_d3d11.c:1271]: Accidental spacing?

**Review by TheSpydog — APPROVED** [review]

**Pull request merged**